editor/src/core/property-controls/property-controls-local.ts

```diff
@@ -74,6 +61,164 @@ async function parseInsertOption(
+    return []
+  }
+  const exportDefaultIdentifier = parseResult.parseResult.exportsDetail.find(isExportDefault)
+  if (exportDefaultIdentifier?.name == null) {
+    // TODO: error handling
+    console.warn('No export default in descriptor file')
+    return []
+  }
+
+  const combinedTopLevelArbitraryBlock = parseResult.parseResult.combinedTopLevelArbitraryBlock
+
+  if (combinedTopLevelArbitraryBlock == null) {
+    // TODO: error handling
+    return []
+  }
+
+  try {
+    // TODO: provide execution scope, so imports can be resolved
+    const evaluatedFile = resolveParamsAndRunJsCode(
+      parseResult.filename,
+      combinedTopLevelArbitraryBlock,
+      {},
+      {},
+    )
+
+    const descriptors = evaluatedFile[exportDefaultIdentifier.name]
+
+    if (descriptors == null) {
+      // TODO: error handling
+      console.warn('Could not find component descriptors in the descriptor file')
+      return []
+    }
+
+    let result: ComponentDescriptorWithName[] = []
+
+    for await (const [moduleName, descriptor] of Object.entries(descriptors)) {
+      const parsedComponents = parseComponents(descriptor)
+
+      if (parsedComponents.type === 'LEFT') {
+        continue
+      }
+
+      for await (const [componentName, componentToRegister] of Object.entries(
+        parsedComponents.value,
+      )) {
+        const componentDescriptor = await componentDescriptorForComponentToRegister(
+          componentToRegister,
+          componentName,
+          moduleName,
+          workers,
+        )
+
+        if (componentDescriptor.type === 'RIGHT') {
+          result = result.concat(componentDescriptor.value)
+        }
+      }
+    }
+    return result
+  } catch {
+    // TODO: error handling
+    console.warn('Error evaluating component descriptor file')
+    return []
+  }
+}
+
+export async function maybeUpdatePropertyControls(
+  parseResults: Array<ParseOrPrintResult>,
+  workers: UtopiaTsWorkers,
+  dispatch: EditorDispatch,
+) {
+  let componentDescriptorUpdates: ComponentDescriptorFileLookup = {}
+
+  const componentDescriptorParseResults = parseResults.filter((p) =>
+    isComponentDescriptorFile(p.filename),
+  )
+  if (componentDescriptorParseResults.length === 0) {
+    // there is nothing to update, return early so no empty dispatch is made
+    return
+  }
+
+  for await (const file of componentDescriptorParseResults) {
+    const descriptors = await getComponentDescriptorPromisesFromParseResult(file, workers)
+    if (descriptors.length > 0) {
+      componentDescriptorUpdates[file.filename] = descriptors
+    }
+  }
+
+  const updatedPropertyControlsInfo = updatePropertyControlsOnDescriptorFileUpdate(
+    componentDescriptorUpdates,
+  )
+  dispatch([EditorActions.updatePropertyControlsInfo(updatedPropertyControlsInfo)])
+}
+
+interface ComponentDescriptorFileLookup {
+  [path: string]: ComponentDescriptorWithName[]
+}
+
+const COMPONENTS_DESCRIPTOR_FILE_CACHE: { current: ComponentDescriptorFileLookup } = {
```

Comment:
(Not for this PR) I'm thinking we should move this into the editor store so that we can avoid globals here (for the sake of tests)